be/btree.c

```diff
@@ -937,21 +1044,42 @@ static int be_btree_delete_key(struct m0_be_btree *tree,
 				break;
 			}
 
-			/*  Retrieve index of the key equal to or greater than*/
-			/*  key being searched */
-			iter = 0;
-			while (iter < bnode->bt_num_active_key &&
-			       key_gt(tree, key,
-				      bnode->bt_kv_arr[iter].btree_key))
-				iter++;
-
-			idx = iter;
-
-			/* check if key is found */
-			if (iter < bnode->bt_num_active_key &&
-			    key_eq(tree, key, bnode->bt_kv_arr[iter].btree_key))
-				break;
-
+			if(tree->bb_ops->ko_type == M0_BBT_CAS_CTG || tree->bb_ops->ko_type == M0_BBT_CONFDB 
+			||tree->bb_ops->ko_type == M0_BBT_INVALID|| tree->bb_ops->ko_type == M0_BBT_UT_KV_OPS
+			|| tree->bb_ops->ko_type == M0_BBT_NR || tree->bb_ops->ko_type == M0_BBT_COB_FILEATTR_EA 
+			|| tree->bb_ops->ko_type == M0_BBT_COB_NAMESPACE ||tree->bb_ops->ko_type == M0_BBT_BALLOC_GROUP_EXTENTS
+			|| tree->bb_ops->ko_type == M0_BBT_BALLOC_GROUP_DESC || tree->bb_ops->ko_type == M0_BBT_EMAP_EM_MAPPING)
+			{
+				/*  Retrieve index of the key equal to or greater than*/
+				/*  key being searched */
+				iter = 0;
+				while (iter < bnode->bt_num_active_key &&
+					key_gt(tree, key,
+						bnode->bt_kv_arr[iter].btree_key))
+					iter++;
+
+				idx = iter;
+
+				/* check if key is found */
+				if (iter < bnode->bt_num_active_key &&
+					key_eq(tree, key, bnode->bt_kv_arr[iter].btree_key))
+					break;
+			}
+			else
+			{
+				iter = 0;
+				while (iter < bnode->bt_num_active_key &&
+					strcmp(strkey, bnode->bt_kv_arr[iter].inlkey) > 0)
+					iter++;
+
+				idx = iter;
+
+				/* check if key is found */
+				if (iter < bnode->bt_num_active_key &&
+					strcmp(strkey, bnode->bt_kv_arr[iter].inlkey) == 0)
+					break;
+			}
```

Comment:
These two blocks of code are almost identical except that one is using key_gt() and another is using strcmp().

Please rewrite function to factor out the common code.

The same applies to other functions above and below.

Reply:
Ok.